mountpoint-s3-crt/src/s3/client.rs

```diff
@@ -792,12 +868,104 @@ impl RequestMetrics {
         unsafe { Some(Headers::from_crt(NonNull::new_unchecked(out))) }
     }
 
-    /// Return the total duration for this request
-    pub fn total_duration(&self) -> Duration {
-        let mut out: u64 = 0;
+    /// Get the IP address the request connected to
+    pub fn request_path_query(&self) -> Option<String> {
+        let mut out: *const aws_string = std::ptr::null();
         // SAFETY: `inner` is a valid aws_s3_request_metrics
-        unsafe { aws_s3_request_metrics_get_total_duration_ns(self.inner.as_ptr(), &mut out) };
-        Duration::from_nanos(out)
+        unsafe {
+            aws_s3_request_metrics_get_request_path_query(self.inner.as_ptr(), &mut out);
+        };
+        if out.is_null() {
+            return None;
+        }
+        // SAFETY: `out` is now a valid pointer to an aws_string, and we'll copy the bytes
+        // out of it so it won't live beyond this function call
+        unsafe {
+            let byte_cursor = aws_byte_cursor_from_string(out);
+            let os_str = OsStr::from_bytes(aws_byte_cursor_as_slice(&byte_cursor));
+            Some(os_str.to_string_lossy().into_owned())
+        }
+    }
+
+    /// Get the host address of the request
+    pub fn host_address(&self) -> Option<String> {
+        let mut out: *const aws_string = std::ptr::null();
+        // SAFETY: `inner` is a valid aws_s3_request_metrics
+        unsafe {
+            aws_s3_request_metrics_get_host_address(self.inner.as_ptr(), &mut out);
+        };
+        if out.is_null() {
+            return None;
+        }
+        // SAFETY: `out` is now a valid pointer to an aws_string, and we'll copy the bytes
+        // out of it so it won't live beyond this function call
+        unsafe {
+            let byte_cursor = aws_byte_cursor_from_string(out);
+            let os_str = OsStr::from_bytes(aws_byte_cursor_as_slice(&byte_cursor));
+            Some(os_str.to_string_lossy().into_owned())
+        }
+    }
+
+    /// Get the IP address the request connected to
+    pub fn ip_address(&self) -> Option<String> {
+        let mut out: *const aws_string = std::ptr::null();
+        // SAFETY: `inner` is a valid aws_s3_request_metrics
+        unsafe {
+            aws_s3_request_metrics_get_ip_address(self.inner.as_ptr(), &mut out)
+                .ok_or_last_error()
+                .ok()?
+        };
+        assert!(!out.is_null(), "IP address should be available if call succeeded");
```

Comment:
Should this be debug assert, and return `None` for release builds?

```suggestion
        debug_assert!(!out.is_null(), "IP address should be available if call succeeded");
        if (out.is_null()) {
            return None;
        }
```